src/mpid/ch3/channels/nemesis/include/mpid_nem_defs.h

```diff
@@ -42,13 +41,14 @@ extern  char *MPID_nem_asymm_base_addr;
 
 static inline MPID_nem_cell_ptr_t MPID_NEM_REL_TO_ABS (MPID_nem_cell_rel_ptr_t r)
 {
-    return (MPID_nem_cell_ptr_t)((char*)MPL_atomic_relaxed_load_ptr(&r.p) + (MPID_nem_addr_t)MPID_nem_asymm_base_addr);
+    void *p = MPL_atomic_relaxed_load_ptr(&r.p);
+    return (MPID_nem_cell_ptr_t) (void *) ((uintptr_t) p + (uintptr_t) MPID_nem_asymm_base_addr);
 }
 
 static inline MPID_nem_cell_rel_ptr_t MPID_NEM_ABS_TO_REL (MPID_nem_cell_ptr_t a)
 {
     MPID_nem_cell_rel_ptr_t ret;
-    MPL_atomic_relaxed_store_ptr(&ret.p, (char *)a - (MPID_nem_addr_t)MPID_nem_asymm_base_addr);
```

Comment:
Couldn't we remove the `MPID_nem_addr_t` cast instead? Then it would be arithmetic on 2 `char` pointers. Or does that still warn?

Reply:
That probably will work. 
But since we used the `uintptr_t` cast in `MPID_NEM_REL_TO_ABS`, I think it is more consistent to use `uintptr_t` here as well. "pointer subtract pointer" is not a common pattern although ligit, it may not improve readability.

Reply:
If we are being *really* pedantic, doing pointer arithmetic by first casting to `uintptr_t` is less safe/portable than doing arithmetic on `char *`. I'm sure MPICH is by no means clean in this regard, but it's something that comes to mind when I see these kind of casts. https://stackoverflow.com/a/43419534/2748723

Reply:
I commented on this instances, but I think it applies to both cases. Just remove the `MPID_nem_addr_t` cast and you are doing `char *` arithmetic.

Reply:
Also, even in this case, the type of `char *` subtract `char *` is `ptrdiff_t` and then we are casting `ptrdiff_t` to a pointer, which is equally undefined behavior. Therefore, I think casting to `uintptr_t` at least the arithmetic part has more semantic sense than pointer arithmetics.

Reply:
Also re-reading this SO comment https://stackoverflow.com/a/43419534/2748723, it was saying that casting to `uintptr_t` and then doing arithmetic is *integer arithmetics*. Note here we actually mean exactly *integer arithmetics*. The difference between "ABS" address and "REL" address is calculated as integer difference, not pointer difference which take into account of pointer type. We are using and treating it as `void *`, it doesn't not have a specific pointer type, thus pointer arithmetics are not *pedanticly* well defined here.

Reply:
I was thinking specifically about this passage.
```
byte pointers (void*, char*) contained a word address with a 3-bit offset stored in the otherwise unused high-order 3 bits of the 64-bit word. Pointer/integer conversions simply copied the representation. The result was that adding 1 to a char* pointer would cause it to point to the next byte (with the offset handled in software), but converting to uintptr_t and adding 1 would cause it to point to the next word.
```
To me, it is saying that if you are calculating byte offsets, `char *` is the only safe way to do so. Again, I think encountering a platform like this is pretty unlikely these days.

Reply:
Pedantically, address arithmetics only defined in the semantics of array indexing. It is undefined if the pointer does not point to the members of the same array. Thus casting to `char *` then doing arithmetic the using the resulting address as a pointer to a non-char array, is equally undefined.

I remember the real issue now -- let's put this pedantic debate aside for a moment. in `pointer +/- number`, the latter is `ptrdiff_t`, which is *signed*. And in our case the `base_addr` can overflow and become negative on 32-bit platform. That is what `ubsan` is complaining. Evan when we cast the `base_addr` to `uintptr_t`, internally it is still casting to `ptrdiff_t` as required by the C sematics and causing the same `ubsan` warning. The only solution is to do interger arithmetics instead.

Reply:
The relative address should be the one stored as `ptrdiff_t`, right? Then it could be added to the base address.

Reply:
Not always. In the case of non-symmetric allocation, the `base_addr` is 0 and the relative address is the actual address (that will overflow `ptrdiff_t`).